diff --git a/vendor/magento/module-inventory-source-deduction-api/Model/SourceDeductionService.php b/vendor/magento/module-inventory-source-deduction-api/Model/SourceDeductionService.php
index f837fe2..41dd600 100644
--- a/vendor/magento/module-inventory-source-deduction-api/Model/SourceDeductionService.php
+++ b/vendor/magento/module-inventory-source-deduction-api/Model/SourceDeductionService.php
@@ -10,6 +10,7 @@ namespace Magento\InventorySourceDeductionApi\Model;
 use Magento\Framework\Exception\LocalizedException;
 use Magento\Inventory\Model\SourceItem\Command\DecrementSourceItemQty;
 use Magento\InventoryApi\Api\Data\SourceItemInterface;
+use Magento\InventoryApi\Api\SourceItemsSaveInterface;
 use Magento\InventoryConfigurationApi\Api\Data\StockItemConfigurationInterface;
 use Magento\InventoryConfigurationApi\Api\GetStockItemConfigurationInterface;
 use Magento\InventorySalesApi\Api\GetStockBySalesChannelInterface;
@@ -44,6 +45,11 @@ class SourceDeductionService implements SourceDeductionServiceInterface
      */
     private $decrementSourceItem;
 
+    /**
+     * @var SourceItemsSaveInterface
+     */
+    private SourceItemsSaveInterface $_sourceItemsSave;
+
     /**
      * @param GetSourceItemBySourceCodeAndSku $getSourceItemBySourceCodeAndSku
      * @param GetStockItemConfigurationInterface $getStockItemConfiguration
@@ -54,12 +60,14 @@ class SourceDeductionService implements SourceDeductionServiceInterface
         GetSourceItemBySourceCodeAndSku $getSourceItemBySourceCodeAndSku,
         GetStockItemConfigurationInterface $getStockItemConfiguration,
         GetStockBySalesChannelInterface $getStockBySalesChannel,
-        DecrementSourceItemQty $decrementSourceItem
+        DecrementSourceItemQty $decrementSourceItem,
+        SourceItemsSaveInterface $sourceItemsSave
     ) {
         $this->getSourceItemBySourceCodeAndSku = $getSourceItemBySourceCodeAndSku;
         $this->getStockItemConfiguration = $getStockItemConfiguration;
         $this->getStockBySalesChannel = $getStockBySalesChannel;
         $this->decrementSourceItem = $decrementSourceItem;
+        $this->_sourceItemsSave = $sourceItemsSave;
     }
 
     /**
@@ -92,10 +100,16 @@ class SourceDeductionService implements SourceDeductionServiceInterface
                     $sourceItem
                 );
                 $sourceItem->setStatus($stockStatus);
-                $sourceItemDecrementData[] = [
-                    'source_item' => $sourceItem,
-                    'qty_to_decrement' => $qty
-                ];
+                $sourceItems[] = $sourceItem;
+            } elseif (($sourceItem->getQuantity() - $qty) < 0) {
+                // If it's a negative inventory, it sets qty to 0.
+                $sourceItem->setQuantity(0);
+                $stockStatus = $this->getSourceStockStatus(
+                    $stockItemConfiguration,
+                    $sourceItem
+                );
+                $sourceItem->setStatus($stockStatus);
+                $sourceItems[] = $sourceItem;
             } else {
                 throw new LocalizedException(
                     __('Not all of your products are available in the requested quantity.')
@@ -103,8 +117,8 @@ class SourceDeductionService implements SourceDeductionServiceInterface
             }
         }
 
-        if (!empty($sourceItemDecrementData)) {
-            $this->decrementSourceItem->execute($sourceItemDecrementData);
+        if (!empty($sourceItems)) {
+            $this->_sourceItemsSave->execute($sourceItems);
         }
     }
 
